src/rules/maxNestingDepthRule.ts

```diff
@@ -0,0 +1,183 @@
+            and decrease each time a control statement is entered and left in the function until its
+            threshold is reached.`,
+        rationale: Lint.Utils.dedent`
+            Maximum nesting depth is a code metric which indicates how deep control statements are
+            nested in a function. This metric quickly reveals the functions whose code is made
+            difficult to read by an excessive nesting of the control statements.
+
+            It's better to have smaller, single-purpose functions with self-documenting names.`,
+        optionsDescription: Lint.Utils.dedent`
+            An optional upper limit for maximum nesting depth can be specified. If no limit option
+            is provided a default value of ${Rule.DEFAULT_THRESHOLD} will be used.`,
+        options: {
+            type: "number",
+            minimum: Rule.MINIMUM_THRESHOLD,
+        },
+        optionExamples: [true, [true, Rule.THRESHOLD_EXAMPLE]],
+        type: "maintainability",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    /**
+     * Builds and returns the failure string
+     * @param expected Rule threshold
+     * @param actual Depth of current function
+     * @param name Function name
+     */
+    public static FAILURE_STRING(expected: number, actual: number, name?: string): string {
+        return (
+            `The function${name === "" ? "" : ` ${name}`} has a nesting depth of at least ` +
+            `${actual} which is higher than the threshold of ${expected}`
+        );
+    }
+
+    /**
+     *  Returns the provided function's name
+     * @param node The function
+     */
+    public static getFunctionName(node: ts.FunctionLikeDeclaration): string {
+        return node.name !== undefined && isIdentifier(node.name) ? node.name.text : "";
+    }
+
+    /**
+     * Returns true if the nesting depth rule is applicable to the current node
+     */
+    public static isNestingDepthStatement(node: ts.Node): boolean {
+        switch (node.kind) {
+            case ts.SyntaxKind.ConditionalExpression:
+            case ts.SyntaxKind.IfStatement:
+            case ts.SyntaxKind.SwitchStatement:
+            case ts.SyntaxKind.DoStatement:
+            case ts.SyntaxKind.WhileStatement:
+            case ts.SyntaxKind.ForStatement:
+            case ts.SyntaxKind.ForInStatement:
+            case ts.SyntaxKind.ForOfStatement:
+            case ts.SyntaxKind.TryStatement:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    /**
+     * Walks the AST
+     * @param ctx Walk context
+     */
+    public static walk(ctx: Lint.WalkContext<{ threshold: number }>): void {
+        const {
+            options: { threshold },
+        } = ctx;
+        const nestingDepthStack: number[] = [];
+        const functionNameStack: string[] = [];
+        const cbNode: (node: ts.Node) => void = (node: ts.Node): void => {
```

Comment:
Nit: the `: (node: ts.Node) => void` typedef seems unnecessary and makes it a little harder to read through this function. Let's remove it?

```suggestion
        const cbNode = (node: ts.Node): void => {
```